src/bosh-director/lib/bosh/director/problem_resolver.rb

```diff
@@ -52,16 +60,65 @@ def apply_resolutions(resolutions)
 
     private
 
-    def process_problem(problem)
-      if problem.state != 'open'
-        reason = "state is '#{problem.state}'"
-        track_and_log("Ignoring problem #{problem.id} (#{reason})")
-      elsif problem.deployment_id != @deployment.id
-        reason = 'not a part of this deployment'
-        track_and_log("Ignoring problem #{problem.id} (#{reason})")
+    def parallel_each(n_threads, ary)
+      if n_threads > 1
+        ThreadPool.new(max_threads: n_threads).wrap do |pool|
+          ary.each do |entry|
+            pool.process do
+              yield entry
+            end
+          end
+        end
       else
+        ary.each do |entry|
+          yield entry
+        end
+      end
+    end
+
+    def process_problem(problem)
+      if problem.open?
         apply_resolution(problem)
+      else
+        track_and_log("Ignoring problem #{problem.id} (state is '#{problem.state}')")
+      end
+    end
+
+    def problems_serially_ordered_by_job(ig_to_problems, &block)
+      BatchMultiInstanceGroupUpdater.partition_jobs_by_serial(@instance_groups).each do |jp|
+        igs_with_problems = []
+        jp.each { |ig| igs_with_problems << ig.name if ig_to_problems.key?(ig.name) }
+        n_threads = [igs_with_problems.size, Config.max_threads].min
+        parallel_each(jp.first.update.serial? ? 1 : n_threads, igs_with_problems) do |ig_name|
+          process_ig(ig_name, ig_to_problems[ig_name], block)
+        end
```

Comment:
This block feels very hard to read, what does `jp` represent? There's plenty of work being done within the iterators in this block. For the sake of readability, do you mind refactoring the variable names and expanding the blocks?